website/scripts/ros/ros_traffic_signal.js

```diff
@@ -30,119 +30,239 @@ function TrafficSignalInfoList(){
+                                        case TRAFFIC_SIGNAL_PHASE_STATE.protected_clearance:
+                                            $('.traffic-signal-col').html(updateTrafficSignal('yellow',remaining_time));
+                                            break;
+                                        case TRAFFIC_SIGNAL_PHASE_STATE.permissive_movement_allowed:
+                                            $('.traffic-signal-col').html(updateTrafficSignal('flash_green',remaining_time));
+                                            break;
+                                        case TRAFFIC_SIGNAL_PHASE_STATE.permissive_clearance:
+                                        case TRAFFIC_SIGNAL_PHASE_STATE.caution_conflicting_traffic:
+                                            $('.traffic-signal-col').html(updateTrafficSignal('flash_yellow',remaining_time));
+                                            break;
+                                        case TRAFFIC_SIGNAL_PHASE_STATE.stop_then_proceed:
+                                            $('.traffic-signal-col').html(updateTrafficSignal('flash_red',remaining_time));
+                                            break;
+                                        default:
+                                            $('.traffic-signal-col').html(updateTrafficSignal('',''));
+                                            console.error("Traffic signal state is invalid");
+                                            break;
+                                    }
+                                    return false;
+                                });
                             }
-                            
                         });
-                    }                    
+                    }
                 });
             }
-         }catch(error){
+        } catch(error) {
             console.error(error);
-         }  
+        }
 
-         //set back to black after 1 second.
-        setInterval(function()
-        {
-            let elapsed_time = Date.now() - latest_start_time;
-            if(elapsed_time >= 1000)
-            {
+        // Create only one interval to avoid memory leaks
+        if (!window.signalIntervalCreated) {
+            window.signalIntervalCreated = true;
+            setInterval(function() {
+                let elapsed_time;
+
+                // Use sim clock for timing if available
+                if (useSimClock && simClockTime !== null && sim_start_time !== null) {
+                    // Calculate elapsed time in simulation
+                    let currentRealTime = Date.now();
+                    let realTimeDiff = currentRealTime - lastRealTime;
+
+                    // Estimate current sim time based on real time elapsed and last known sim time
+                    // This handles the case where sim_clock messages come at lower frequency than needed
+                    let estimatedSimTime = lastSimTime + realTimeDiff;
```

Comment:
Not sure how these real and sim time get twisted together.

Reply:
Sorry, this was a mistake. Added a cleaner code